genai-perf/genai_perf/checkpoint/checkpoint.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import os
+from dataclasses import dataclass
+
+from genai_perf.config.input.config_command import ConfigCommand
+from genai_perf.config.run.results import Results
+from genai_perf.exceptions import GenAIPerfException
+from genai_perf.types import CheckpointObject
+
+
+@dataclass(frozen=True)
```

Comment:
This is how I've decided to do constants (after consulting w/ Elias). I feel like it's a really clean way to declare constants and makes importing them trivially easy (as you don't need to declare each constant individually).

I would encourage us to use this going forward throughout GAP.

Reply:
I see. What are your thoughts, @nicolasnoble @matthewkotila?

I'm open to any approach for Python constants. I don't have time today to research the best practice for declaring constants in Python. I would just like to make sure we're not developing with two different approaches in our codebase.

Reply:
Why not just have the constant as a data member in the class itself? Looks like it's not used anywhere else currently.

Reply:
See the example in `ConfigCommand` for a more extensive usage of this w/ RunConfigDefaults.

Reply:
Seems like RunConfigDefaults could be a separate file with non-class-data-member constants. That's the more typical/idiomatic way to do constants in Python rather than making a class and setting data members.

If you want to import a bunch of stuff from a run_config_defaults.py you can do `import run_config_defaults` and then use them like `run_config_defaults.CHECKPOINT_DIRECTORY` or `from run_config_defaults import *` then `CHECKPOINT_DIRECTORY`

Reply:
We can't import w/ a wildcard and it's onerous to import all the constants you need. I thought this was a pretty nifty solution compared to what we did in MA.

I'm failing to understand the downside of coding it this way.

Reply:
The main downside I see is that it's atypical, which makes maintaining it harder.

Why can't we import with wildcard?

Reply:
If there were a way to say "I feel like 3/10 strength for this discussion, but I'm having it anyway in case either of us learns something", I would haha.

Reply:
Wildcard imports are generally considered very bad, as they pollute the namespace.

I spent some time looking into this. The current approach in GenAI-Perf of having a dedicated constants module (or multiple for grouped constants) is one best practice. Another best practice is using namespace classes. Frozen dataclasses are also a best practice. They add immutability and structure, but namespace classes are sometimes preferred if you don't need all of the functions that dataclasses generate. They all haves pros and cons, so it's a choice of what's best for the codebase.

I don't particularly mind which approach we go with, but we should select one approach before this code is added into GAP. If it's frozen dataclasses, then we should redo the GAP code to change to that approach. We should generally make sure that we're following similar practices across the codebase. The feature branch is going to be really big, so if we start to deviate in coding practices between this feature branch and the rest of the codebase, we should find a way to reconcile them into one codebase with one set of practices before merging.

Reply:
I'll resolve this conversation, as Brian is trying to get this in before his vacation. We'll figure out how to ensure we're consistent with coding practices in the future. :)